program-runtime/src/loaded_programs.rs

```diff
@@ -914,6 +952,39 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
         self.unload_program_entries(sorted_candidates.iter().take(num_to_unload));
     }
 
+    /// Evicts programs using 2's random selection, choosing the least used program out of the two entries.
+    /// The eviction is performed enough number of times to reduce the cache usage to the given percentage.
+    pub fn evict_using_2s_random_selection(&mut self, shrink_to: PercentageInteger, now: Slot) {
+        let mut candidates = self.get_flattened_entries();
+        let num_to_unload = candidates
+            .len()
+            .saturating_sub(shrink_to.apply_to(MAX_LOADED_ENTRY_COUNT));
+        for _ in 0..num_to_unload {
+            let mut rng = thread_rng();
+            let index1 = rng.gen_range(0..candidates.len());
+            let index2 = rng.gen_range(0..candidates.len());
+            if let Some((usage_counter1, usage_counter2)) =
+                candidates.get(index1).and_then(|(_id, entry1)| {
+                    candidates.get(index2).map(|(_id, entry2)| {
+                        (
+                            entry1.decayed_usage_counter(now),
+                            entry2.decayed_usage_counter(now),
+                        )
+                    })
+                })
+            {
+                let (program, entry) = if usage_counter1 < usage_counter2 {
+                    candidates.remove(index1)
+                } else {
+                    candidates.remove(index2)
+                };
+                self.unload_program_entry(&program, &entry);
+            } else {
+                error!("Failed in evicting an entry from the program cache");
```

Comment:
in this case the cache is growing without bound?

when could the candidates be `None`?

Reply:
It should never be `None`, since the random index is within the range. I didn't want to do `unwrap()` or `expect()`. What would be a semantically better approach?